src/services/Carts.php

```diff
@@ -537,4 +538,27 @@ private function _getCartPaymentCurrencyIso(): string
 
         return Plugin::getInstance()->getPaymentCurrencies()->getPrimaryPaymentCurrencyIso();
     }
+
+    /**
+     * @param User $user
+     * @return void
+     * @throws MissingComponentException
+     * @throws Throwable
+     */
+    public function customerRegisteredHandler(User $user): void
+    {
+        $segments = Craft::$app->getRequest()->getActionSegments();
+        $userSaveSegments = ['users', 'save-user'];
+        $isUserSaveAction = $segments == $userSaveSegments;
+
+        // we have a cart number, currently anon, and the current action being executed is user save
+        if (!Craft::$app->getUser()->getIdentity() &&
+            !Craft::$app->getRequest()->getIsCpRequest() &&
+            $isUserSaveAction
+        ) {
+            $currentCartNumber = $this->getSessionCartNumber();
+            // Set the session flag to preserve the cart for this user
+            Craft::$app->getSession()->set('commerce:anonymousCartWithCredentialedCustomer:' . $currentCartNumber, true);
```

Comment:
This key is used in a few places now, should we make a common method or constant to retrieve it?